agent/consul/stream/event.go

```diff
@@ -19,71 +23,70 @@ type Event struct {
 }
 
 type Payload interface {
-	// FilterByKey must return true if the Payload should be included in a subscription
+	// MatchesKey must return true if the Payload should be included in a subscription
 	// requested with the key and namespace.
 	// Generally this means that the payload matches the key and namespace or
 	// the payload is a special framing event that should be returned to every
 	// subscription.
-	FilterByKey(key, namespace string) bool
+	MatchesKey(key, namespace string) bool
+
+	// HasReadPermission uses the acl.Authorizer to determine if the items in the
+	// Payload are visible to the request. It returns true if the payload is
+	// authorized for Read, otherwise returns false.
+	HasReadPermission(authz acl.Authorizer) bool
 }
 
-// Len returns the number of events contained within this event. If the Payload
-// is a []Event, the length of that slice is returned. Otherwise 1 is returned.
-func (e Event) Len() int {
-	if batch, ok := e.Payload.(PayloadEvents); ok {
-		return len(batch)
-	}
-	return 1
+// PayloadEvents is an Payload which contains multiple Events.
+type PayloadEvents struct {
+	Items []Event
 }
 
-// Filter returns an Event filtered to only those Events where f returns true.
-// If the second return value is false, every Event was removed by the filter.
-func (e Event) Filter(f func(Event) bool) (Event, bool) {
-	batch, ok := e.Payload.(PayloadEvents)
-	if !ok {
-		return e, f(e)
-	}
+func NewPayloadEvents(items ...Event) *PayloadEvents {
+	return &PayloadEvents{Items: items}
+}
+
+func (p *PayloadEvents) filter(f func(Event) bool) bool {
+	items := p.Items
 
 	// To avoid extra allocations, iterate over the list of events first and
 	// get a count of the total desired size. This trades off some extra cpu
 	// time in the worse case (when not all items match the filter), for
 	// fewer memory allocations.
 	var size int
-	for idx := range batch {
-		if f(batch[idx]) {
+	for idx := range items {
+		if f(items[idx]) {
 			size++
 		}
 	}
-	if len(batch) == size || size == 0 {
-		return e, size != 0
+	if len(items) == size || size == 0 {
+		return size != 0
 	}
 
-	filtered := make(PayloadEvents, 0, size)
-	for idx := range batch {
-		event := batch[idx]
+	filtered := make([]Event, 0, size)
+	for idx := range items {
+		event := items[idx]
 		if f(event) {
 			filtered = append(filtered, event)
 		}
 	}
-	if len(filtered) == 0 {
-		return e, false
-	}
-	e.Payload = filtered
-	return e, true
+	p.Items = filtered
+	return true
 }
 
-// PayloadEvents is an Payload which contains multiple Events.
-type PayloadEvents []Event
+func (p *PayloadEvents) MatchesKey(key, namespace string) bool {
```

Comment:
Is it weird that a method which seems like a predicate on the value is actually mutating it destructively here? Ironically, the original name `FilterByKey` would have been a better fit here if the intent is that the event might actually update its contents.

In fact is this incorrect and racey? For example is we stored a `PayloadEvents` event in the buffer itself because we need to broadcast multiple events in one raft entry (which is likely) then when one subscriber filters it, won't it be unexpectedly mutating the shared event in place - both incorrectly filtering the result for other subscribes and being racy since there might be readers (and even other filter calls) from other subscriptions concurrently?

If not, and it's OK because we currently only ever use this in the context of a single subscription, then I think it would be OK for now if we comment it to make it explicit that it can't ever be stored in the shared buffers or snapshots and that this method is destructive.

Reply:
The `PayloadEvents` is created in `Subscription.Next` after we get the slice of events out of the event buffer item. It should never be submitted to `Publish` or used in the event buffers. That is currently not documented, but it should be. I'll update the godoc to include this detail and also that this method does mutate the underlying slice.

I'd also be happy to go back to `FilterByKey` if that is a better fit now.